In [1]:
import pandas as pd

In [2]:
matches = pd.read_csv('matches.csv', index_col=0)

In [3]:
matches.head()

date   time            comp        round  day venue result   gf   ga  \
1  2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2  2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3  2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4  2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6  2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  match report  notes    sh   sot  dist   fk   pk pkatt  \
1       Tottenham  ...  Match Report    NaN  18.0   4.0  16.9  1.0  0.0   0.0   
2    Norwich City  ...  Match Report    NaN  16.0   4.0  17.3  1.0  0.0   0.0   
3         Arsenal  ...  Match Report    NaN  25.0  10.0  14.3  0.0  0.0   0.0   
4  Leicester City  ...  Match Report    NaN  25.0   8.0  14.0  0.0  0.0   0.0   
6     Southampton  ...  Match Report    NaN  16.0   1.0  15.7  1.0  0.0   0.0   

   season             team  
1    2022  Manchester City  
2    2022  Manchester City  
3    2022  Manchester City  
4    2022  Manchester City  
6    2022  Manchester City  

[5 rows x 27 columns]

In [4]:
matches.shape

(1389, 27)

In [5]:
38 * 20 * 2

1520

Clearly there is some missing data as the total number of matches that should have been played in 2 seasons were 1520 but we have data of only 1389 matches.

In [6]:
matches["team"].value_counts()

team
Southampton                 72
Brighton and Hove Albion    72
Manchester United           72
West Ham United             72
Newcastle United            72
Burnley                     71
Leeds United                71
Crystal Palace              71
Manchester City             71
Wolverhampton Wanderers     71
Tottenham Hotspur           71
Arsenal                     71
Leicester City              70
Chelsea                     70
Aston Villa                 70
Everton                     70
Liverpool                   38
Fulham                      38
West Bromwich Albion        38
Sheffield United            38
Brentford                   34
Watford                     33
Norwich City                33
Name: count, dtype: int64

In [7]:
matches['round'].value_counts()

round
Matchweek 1     39
Matchweek 16    39
Matchweek 34    39
Matchweek 32    39
Matchweek 31    39
Matchweek 29    39
Matchweek 28    39
Matchweek 26    39
Matchweek 25    39
Matchweek 24    39
Matchweek 23    39
Matchweek 2     39
Matchweek 19    39
Matchweek 17    39
Matchweek 20    39
Matchweek 15    39
Matchweek 5     39
Matchweek 3     39
Matchweek 13    39
Matchweek 12    39
Matchweek 4     39
Matchweek 11    39
Matchweek 10    39
Matchweek 9     39
Matchweek 8     39
Matchweek 14    39
Matchweek 7     39
Matchweek 6     39
Matchweek 30    37
Matchweek 27    37
Matchweek 22    37
Matchweek 21    37
Matchweek 18    37
Matchweek 33    32
Matchweek 35    20
Matchweek 36    20
Matchweek 37    20
Matchweek 38    20
Name: count, dtype: int64

In [8]:
matches.dtypes

date             object
time             object
comp             object
round            object
day              object
venue            object
result           object
gf              float64
ga              float64
opponent         object
xg              float64
xga             float64
poss            float64
attendance      float64
captain          object
formation        object
referee          object
match report     object
notes           float64
sh              float64
sot             float64
dist            float64
fk              float64
pk              float64
pkatt           float64
season            int64
team             object
dtype: object

In [9]:
matches["date"] = pd.to_datetime(matches["date"])

In [10]:
matches["venue_code"] = matches["venue"].astype("category").cat.codes

In [11]:
matches["opp_code"] = matches["opponent"].astype("category").cat.codes

In [12]:
matches["hours"] = matches["time"].str.replace(":.+", "", regex=True).astype("int")

In [13]:
matches["day_code"] = matches["date"].dt.dayofweek

In [14]:
matches["target"] = (matches["result"] == "W").astype("int")

In [15]:
matches.head()

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...   fk   pk  pkatt  season             team venue_code  \
1       Tottenham  ...  1.0  0.0    0.0    2022  Manchester City          0   
2    Norwich City  ...  1.0  0.0    0.0    2022  Manchester City          1   
3         Arsenal  ...  0.0  0.0    0.0    2022  Manchester City          1   
4  Leicester City  ...  0.0  0.0    0.0    2022  Manchester City          0   
6     Southampton  ...  1.0  0.0    0.0    2022  Manchester City          1   

  opp_code hours  day_code  target  
1       18    16         6       0  
2       15    15         5       1  
3        0    12         5       1  
4       10    15         5       1  
6       17    15         5       0  

[5 rows x 32 columns]

In [16]:
from sklearn.ensemble import RandomForestClassifier

In [17]:
rf = RandomForestClassifier(n_estimators = 50, min_samples_split = 10, random_state = 1)

In [18]:
train = matches[matches["date"] < '2022-01-01']

In [19]:
test = matches[matches["date"] > '2022-01-01']

In [20]:
predictors = ["venue_code", "opp_code", "hours", "day_code"]

In [21]:
rf.fit(train[predictors], train["target"])

RandomForestClassifier(min_samples_split=10, n_estimators=50, random_state=1)

In [22]:
preds = rf.predict(test[predictors])

In [23]:
from sklearn.metrics import accuracy_score

In [24]:
acc = accuracy_score(test["target"], preds)

In [25]:
acc

0.6123188405797102

In [26]:
combined = pd.DataFrame(dict(actual = test["target"], prediction = preds))

In [27]:
pd.crosstab(index = combined["actual"], columns = combined["prediction"])

prediction    0   1
actual             
0           141  31
1            76  28

In [28]:
from sklearn.metrics import precision_score

In [29]:
precision_score(test["target"], preds)

0.4745762711864407

In [30]:
grouped_matches = matches.groupby("team")

In [31]:
group = grouped_matches.get_group("Liverpool")

In [32]:
group

date   time            comp         round  day venue result   gf  \
1  2020-09-12  17:30  Premier League   Matchweek 1  Sat  Home      W  4.0   
2  2020-09-20  16:30  Premier League   Matchweek 2  Sun  Away      W  2.0   
4  2020-09-28  20:00  Premier League   Matchweek 3  Mon  Home      W  3.0   
6  2020-10-04  19:15  Premier League   Matchweek 4  Sun  Away      L  2.0   
7  2020-10-17  12:30  Premier League   Matchweek 5  Sat  Away      D  2.0   
9  2020-10-24  20:00  Premier League   Matchweek 6  Sat  Home      W  2.0   
11 2020-10-31  17:30  Premier League   Matchweek 7  Sat  Home      W  2.0   
13 2020-11-08  16:30  Premier League   Matchweek 8  Sun  Away      D  1.0   
14 2020-11-22  19:15  Premier League   Matchweek 9  Sun  Home      W  3.0   
16 2020-11-28  12:30  Premier League  Matchweek 10  Sat  Away      D  1.0   
18 2020-12-06  19:15  Premier League  Matchweek 11  Sun  Home      W  4.0   
20 2020-12-13  16:30  Premier League  Matchweek 12  Sun  Away      D  1.0   
21 2020-12-16  20:00  Premier League  Matchweek 13  Wed  Home      W  2.0   
22 2020-12-19  12:30  Premier League  Matchweek 14  Sat  Away      W  7.0   
23 2020-12-27  16:30  Premier League  Matchweek 15  Sun  Home      D  1.0   
24 2020-12-30  20:00  Premier League  Matchweek 16  Wed  Away      D  0.0   
25 2021-01-04  20:00  Premier League  Matchweek 17  Mon  Away      L  0.0   
27 2021-01-17  16:30  Premier League  Matchweek 19  Sun  Home      D  0.0   
28 2021-01-21  20:00  Premier League  Matchweek 18  Thu  Home      L  0.0   
30 2021-01-28  20:00  Premier League  Matchweek 20  Thu  Away      W  3.0   
31 2021-01-31  16:30  Premier League  Matchweek 21  Sun  Away      W  3.0   
32 2021-02-03  20:15  Premier League  Matchweek 22  Wed  Home      L  0.0   
33 2021-02-07  16:30  Premier League  Matchweek 23  Sun  Home      L  1.0   
34 2021-02-13  12:30  Premier League  Matchweek 24  Sat  Away      L  1.0   
36 2021-02-20  17:30  Premier League  Matchweek 25  Sat  Home      L  0.0   
37 2021-02-28  19:15  Premier League  Matchweek 26  Sun  Away      W  2.0   
38 2021-03-04  20:15  Premier League  Matchweek 29  Thu  Home      L  0.0   
39 2021-03-07  14:00  Premier League  Matchweek 27  Sun  Home      L  0.0   
41 2021-03-15  20:00  Premier League  Matchweek 28  Mon  Away      W  1.0   
42 2021-04-03  20:00  Premier League  Matchweek 30  Sat  Away      W  3.0   
44 2021-04-10  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
46 2021-04-19  20:00  Premier League  Matchweek 32  Mon  Away      D  1.0   
47 2021-04-24  12:30  Premier League  Matchweek 33  Sat  Home      D  1.0   
48 2021-05-08  20:15  Premier League  Matchweek 35  Sat  Home      W  2.0   
49 2021-05-13  20:15  Premier League  Matchweek 34  Thu  Away      W  4.0   
50 2021-05-16  16:30  Premier League  Matchweek 36  Sun  Away      W  2.0   
51 2021-05-19  20:15  Premier League  Matchweek 37  Wed  Away      W  3.0   
52 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  2.0   

     ga         opponent  ...   fk   pk  pkatt  season       team venue_code  \
1   3.0     Leeds United  ...  0.0  2.0    2.0    2021  Liverpool          1   
2   0.0          Chelsea  ...  1.0  0.0    0.0    2021  Liverpool          0   
4   1.0          Arsenal  ...  0.0  0.0    0.0    2021  Liverpool          1   
6   7.0      Aston Villa  ...  1.0  0.0    0.0    2021  Liverpool          0   
7   2.0          Everton  ...  1.0  0.0    0.0    2021  Liverpool          0   
9   1.0    Sheffield Utd  ...  1.0  0.0    0.0    2021  Liverpool          1   
11  1.0         West Ham  ...  1.0  1.0    1.0    2021  Liverpool          1   
13  1.0  Manchester City  ...  0.0  1.0    1.0    2021  Liverpool          0   
14  0.0   Leicester City  ...  0.0  0.0    0.0    2021  Liverpool          1   
16  1.0         Brighton  ...  0.0  0.0    0.0    2021  Liverpool          0   
18  0.0           Wolves  ...  1.0  0.0    0.0    2021  Liverpool          1   
20  1.0           Fulham  ...  1.0  1.0    1.

In [33]:
def rolling_averages(group, cols, new_cols):
    group = group.sort_values("date")
    rolling_stats = group[cols].rolling(3, closed = "left").mean()
    group[new_cols] = rolling_stats
    group = group.dropna(subset = new_cols)
    return group

In [34]:
cols = ["gf", "ga", "sh", "sot", "dist", "fk", "pk", "pkatt"]
new_cols = [f"{c}_rolling" for c in cols]

In [35]:
new_cols

['gf_rolling',
 'ga_rolling',
 'sh_rolling',
 'sot_rolling',
 'dist_rolling',
 'fk_rolling',
 'pk_rolling',
 'pkatt_rolling']

In [36]:
rolling_averages(group, cols, new_cols)

date   time            comp         round  day venue result   gf  \
6  2020-10-04  19:15  Premier League   Matchweek 4  Sun  Away      L  2.0   
7  2020-10-17  12:30  Premier League   Matchweek 5  Sat  Away      D  2.0   
9  2020-10-24  20:00  Premier League   Matchweek 6  Sat  Home      W  2.0   
11 2020-10-31  17:30  Premier League   Matchweek 7  Sat  Home      W  2.0   
13 2020-11-08  16:30  Premier League   Matchweek 8  Sun  Away      D  1.0   
14 2020-11-22  19:15  Premier League   Matchweek 9  Sun  Home      W  3.0   
16 2020-11-28  12:30  Premier League  Matchweek 10  Sat  Away      D  1.0   
18 2020-12-06  19:15  Premier League  Matchweek 11  Sun  Home      W  4.0   
20 2020-12-13  16:30  Premier League  Matchweek 12  Sun  Away      D  1.0   
21 2020-12-16  20:00  Premier League  Matchweek 13  Wed  Home      W  2.0   
22 2020-12-19  12:30  Premier League  Matchweek 14  Sat  Away      W  7.0   
23 2020-12-27  16:30  Premier League  Matchweek 15  Sun  Home      D  1.0   
24 2020-12-30  20:00  Premier League  Matchweek 16  Wed  Away      D  0.0   
25 2021-01-04  20:00  Premier League  Matchweek 17  Mon  Away      L  0.0   
27 2021-01-17  16:30  Premier League  Matchweek 19  Sun  Home      D  0.0   
28 2021-01-21  20:00  Premier League  Matchweek 18  Thu  Home      L  0.0   
30 2021-01-28  20:00  Premier League  Matchweek 20  Thu  Away      W  3.0   
31 2021-01-31  16:30  Premier League  Matchweek 21  Sun  Away      W  3.0   
32 2021-02-03  20:15  Premier League  Matchweek 22  Wed  Home      L  0.0   
33 2021-02-07  16:30  Premier League  Matchweek 23  Sun  Home      L  1.0   
34 2021-02-13  12:30  Premier League  Matchweek 24  Sat  Away      L  1.0   
36 2021-02-20  17:30  Premier League  Matchweek 25  Sat  Home      L  0.0   
37 2021-02-28  19:15  Premier League  Matchweek 26  Sun  Away      W  2.0   
38 2021-03-04  20:15  Premier League  Matchweek 29  Thu  Home      L  0.0   
39 2021-03-07  14:00  Premier League  Matchweek 27  Sun  Home      L  0.0   
41 2021-03-15  20:00  Premier League  Matchweek 28  Mon  Away      W  1.0   
42 2021-04-03  20:00  Premier League  Matchweek 30  Sat  Away      W  3.0   
44 2021-04-10  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
46 2021-04-19  20:00  Premier League  Matchweek 32  Mon  Away      D  1.0   
47 2021-04-24  12:30  Premier League  Matchweek 33  Sat  Home      D  1.0   
48 2021-05-08  20:15  Premier League  Matchweek 35  Sat  Home      W  2.0   
49 2021-05-13  20:15  Premier League  Matchweek 34  Thu  Away      W  4.0   
50 2021-05-16  16:30  Premier League  Matchweek 36  Sun  Away      W  2.0   
51 2021-05-19  20:15  Premier League  Matchweek 37  Wed  Away      W  3.0   
52 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  2.0   

     ga         opponent  ...  day_code  target  gf_rolling  ga_rolling  \
6   7.0      Aston Villa  ...         6       0    3.000000    1.333333   
7   2.0          Everton  ...         5       0    2.333333    2.666667   
9   1.0    Sheffield Utd  ...         5       1    2.333333    3.333333   
11  1.0         West Ham  ...         5       1    2.000000    3.333333   
13  1.0  Manchester City  ...         6       0    2.000000    1.333333   
14  0.0   Leicester City  ...         6       1    1.666667    1.000000   
16  1.0         Brighton  ...         5       0    2.000000    0.666667   
18  0.0           Wolves  ...         6       1    1.666667    0.666667   
20  1.0           Fulham  ...         6       0    2.666667    0.333333   
21  1.0        Tottenham  ...         2       1    2.000000    0.666667   
22  0.0   Crystal Palace  ...         5       1    2.333333    0.666667   
23  1.0        West Brom  ...         6       0    3.333333    0.666667   
24  0.0    Newcastle Utd  ...         2       0    3.333333    0.666667   
25  1.0      Southampton  ...         0       0    2.666667    0.333333   
27  0.0   Manchester Utd  ...         6       0    0.333333    0.666667   
28  1.0          Burnley  ...       

In [37]:
matches_rolling = matches.groupby("team").apply(lambda x: rolling_averages(x, cols, new_cols))

<ipython-input-37-d7a196cb9ce7>:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  matches_rolling = matches.groupby("team").apply(lambda x: rolling_averages(x, cols, new_cols))


In [38]:
matches_rolling

date   time            comp         round  \
team                                                                         
Arsenal                 6  2020-10-04  14:00  Premier League   Matchweek 4   
                        7  2020-10-17  17:30  Premier League   Matchweek 5   
                        9  2020-10-25  19:15  Premier League   Matchweek 6   
                        11 2020-11-01  16:30  Premier League   Matchweek 7   
                        13 2020-11-08  19:15  Premier League   Matchweek 8   
...                               ...    ...             ...           ...   
Wolverhampton Wanderers 32 2022-03-13  14:00  Premier League  Matchweek 29   
                        33 2022-03-18  20:00  Premier League  Matchweek 30   
                        34 2022-04-02  15:00  Premier League  Matchweek 31   
                        35 2022-04-08  20:00  Premier League  Matchweek 32   
                        36 2022-04-24  14:00  Premier League  Matchweek 34   

                            day venue result   gf   ga         opponent  ...  \
team                                                                     ...   
Arsenal                 6   Sun  Home      W  2.0  1.0    Sheffield Utd  ...   
                        7   Sat  Away      L  0.0  1.0  Manchester City  ...   
                        9   Sun  Home      L  0.0  1.0   Leicester City  ...   
                        11  Sun  Away      W  1.0  0.0   Manchester Utd  ...   
                        13  Sun  Home      L  0.0  3.0      Aston Villa  ...   
...                         ...   ...    ...  ...  ...              ...  ...   
Wolverhampton Wanderers 32  Sun  Away      W  1.0  0.0          Everton  ...   
                        33  Fri  Home      L  2.0  3.0     Leeds United  ...   
                        34  Sat  Home      W  2.0  1.0      Aston Villa  ...   
                        35  Fri  Away      L  0.0  1.0    Newcastle Utd  ...   
                        36  Sun  Away      L  0.0  1.0          Burnley  ...   

                            day_code  target  gf_rolling  ga_rolling  \
team                                                                   
Arsenal                 6          6       1    2.000000    1.333333   
                        7          5       0    1.666667    1.666667   
                        9          6       0    1.000000    1.666667   
                        11         6       1    0.666667    1.000000   
                        13         6       0    0.333333    0.666667   
...                              ...     ...         ...         ...   
Wolverhampton Wanderers 32         6       1    1.333333    1.000000   
                        33         4       0    1.666667    0.666667   
                        34         5       1    2.333333    1.000000   
                        35         4       0    1.666667    1.333333   
                        36         6       0    1.333333    1.666667   

                           sh_rolling sot_rolling dist_rolling fk_rolling  \
team                                                                        
Arsenal                 6    7.666667    3.666667    14.733333   0.666667   
                        7    5.333333    3.666667    15.766667   0.000000   
                        9    7.000000    3.666667    16.733333   0.666667   
                        11   9.666667    4.000000    16.033333   1.000000   
                        13   9.666667    2.666667    18.033333   1.000000   
...                               ...         ...          ...        ...   
Wolverhampton Wanderers 32  12.333333    3.666667    19.300000   0.000000   
                        33  12.333333    4.333333    19.600000   0.000000   
                        34  13.000000    5.333333    19.833333   0.000000   
                        35  13.000000    5.000000    18.533333   0.000000   
                        36  10.000000    4.666667    17.633333   0.000000   

                            pk_rolling

In [39]:
matches_rolling = matches_rolling.droplevel('team')

In [40]:
matches_rolling

date   time            comp         round  day venue result   gf  \
6  2020-10-04  14:00  Premier League   Matchweek 4  Sun  Home      W  2.0   
7  2020-10-17  17:30  Premier League   Matchweek 5  Sat  Away      L  0.0   
9  2020-10-25  19:15  Premier League   Matchweek 6  Sun  Home      L  0.0   
11 2020-11-01  16:30  Premier League   Matchweek 7  Sun  Away      W  1.0   
13 2020-11-08  19:15  Premier League   Matchweek 8  Sun  Home      L  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
32 2022-03-13  14:00  Premier League  Matchweek 29  Sun  Away      W  1.0   
33 2022-03-18  20:00  Premier League  Matchweek 30  Fri  Home      L  2.0   
34 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
35 2022-04-08  20:00  Premier League  Matchweek 32  Fri  Away      L  0.0   
36 2022-04-24  14:00  Premier League  Matchweek 34  Sun  Away      L  0.0   

     ga         opponent  ...  day_code  target  gf_rolling  ga_rolling  \
6   1.0    Sheffield Utd  ...         6       1    2.000000    1.333333   
7   1.0  Manchester City  ...         5       0    1.666667    1.666667   
9   1.0   Leicester City  ...         6       0    1.000000    1.666667   
11  0.0   Manchester Utd  ...         6       1    0.666667    1.000000   
13  3.0      Aston Villa  ...         6       0    0.333333    0.666667   
..  ...              ...  ...       ...     ...         ...         ...   
32  0.0          Everton  ...         6       1    1.333333    1.000000   
33  3.0     Leeds United  ...         4       0    1.666667    0.666667   
34  1.0      Aston Villa  ...         5       1    2.333333    1.000000   
35  1.0    Newcastle Utd  ...         4       0    1.666667    1.333333   
36  1.0          Burnley  ...         6       0    1.333333    1.666667   

   sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
6    7.666667    3.666667    14.733333   0.666667    0.000000       0.000000  
7    5.333333    3.666667    15.766667   0.000000    0.000000       0.000000  
9    7.000000    3.666667    16.733333   0.666667    0.000000       0.000000  
11   9.666667    4.000000    16.033333   1.000000    0.000000       0.000000  
13   9.666667    2.666667    18.033333   1.000000    0.333333       0.333333  
..        ...         ...          ...        ...         ...            ...  
32  12.333333    3.666667    19.300000   0.000000    0.000000       0.000000  
33  12.333333    4.333333    19.600000   0.000000    0.000000       0.000000  
34  13.000000    5.333333    19.833333   0.000000    0.000000       0.000000  
35  13.000000    5.000000    18.533333   0.000000    0.000000       0.000000  
36  10.000000    4.666667    17.633333   0.000000    0.000000       0.000000  

[1317 rows x 40 columns]

In [41]:
matches_rolling.index = range(matches_rolling.shape[0])

In [42]:
matches_rolling

date   time            comp         round  day venue result   gf  \
0    2020-10-04  14:00  Premier League   Matchweek 4  Sun  Home      W  2.0   
1    2020-10-17  17:30  Premier League   Matchweek 5  Sat  Away      L  0.0   
2    2020-10-25  19:15  Premier League   Matchweek 6  Sun  Home      L  0.0   
3    2020-11-01  16:30  Premier League   Matchweek 7  Sun  Away      W  1.0   
4    2020-11-08  19:15  Premier League   Matchweek 8  Sun  Home      L  0.0   
...         ...    ...             ...           ...  ...   ...    ...  ...   
1312 2022-03-13  14:00  Premier League  Matchweek 29  Sun  Away      W  1.0   
1313 2022-03-18  20:00  Premier League  Matchweek 30  Fri  Home      L  2.0   
1314 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
1315 2022-04-08  20:00  Premier League  Matchweek 32  Fri  Away      L  0.0   
1316 2022-04-24  14:00  Premier League  Matchweek 34  Sun  Away      L  0.0   

       ga         opponent  ...  day_code  target  gf_rolling  ga_rolling  \
0     1.0    Sheffield Utd  ...         6       1    2.000000    1.333333   
1     1.0  Manchester City  ...         5       0    1.666667    1.666667   
2     1.0   Leicester City  ...         6       0    1.000000    1.666667   
3     0.0   Manchester Utd  ...         6       1    0.666667    1.000000   
4     3.0      Aston Villa  ...         6       0    0.333333    0.666667   
...   ...              ...  ...       ...     ...         ...         ...   
1312  0.0          Everton  ...         6       1    1.333333    1.000000   
1313  3.0     Leeds United  ...         4       0    1.666667    0.666667   
1314  1.0      Aston Villa  ...         5       1    2.333333    1.000000   
1315  1.0    Newcastle Utd  ...         4       0    1.666667    1.333333   
1316  1.0          Burnley  ...         6       0    1.333333    1.666667   

     sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
0      7.666667    3.666667    14.733333   0.666667    0.000000       0.000000  
1      5.333333    3.666667    15.766667   0.000000    0.000000       0.000000  
2      7.000000    3.666667    16.733333   0.666667    0.000000       0.000000  
3      9.666667    4.000000    16.033333   1.000000    0.000000       0.000000  
4      9.666667    2.666667    18.033333   1.000000    0.333333       0.333333  
...         ...         ...          ...        ...         ...            ...  
1312  12.333333    3.666667    19.300000   0.000000    0.000000       0.000000  
1313  12.333333    4.333333    19.600000   0.000000    0.000000       0.000000  
1314  13.000000    5.333333    19.833333   0.000000    0.000000       0.000000  
1315  13.000000    5.000000    18.533333   0.000000    0.000000       0.000000  
1316  10.000000    4.666667    17.633333   0.000000    0.000000       0.000000  

[1317 rows x 40 columns]

In [43]:
def make_prediction(data, predictors):
  train = data[data["date"] < '2022-01-01']
  test = data[data["date"] > '2022-01-01']
  rf.fit(train[predictors], train["target"])
  preds = rf.predict(test[predictors])
  combined = pd.DataFrame(dict(actual = test["target"], prediction = preds), index = test.index)
  precision = precision_score(test["target"], preds)
  return combined, precision

In [44]:
combined, precision = make_prediction(matches_rolling, predictors + new_cols)

In [45]:
precision

0.625

In [46]:
combined

actual  prediction
55         0           0
56         1           0
57         1           0
58         1           1
59         1           1
...      ...         ...
1312       1           0
1313       0           0
1314       1           0
1315       0           0
1316       0           0

[276 rows x 2 columns]

In [47]:
combined = combined.merge(matches_rolling[["date", "team", "opponent", "result"]], left_index = True, right_index = True)

In [48]:
combined

actual  prediction       date                     team       opponent  \
55         0           0 2022-01-23                  Arsenal        Burnley   
56         1           0 2022-02-10                  Arsenal         Wolves   
57         1           0 2022-02-19                  Arsenal      Brentford   
58         1           1 2022-02-24                  Arsenal         Wolves   
59         1           1 2022-03-06                  Arsenal        Watford   
...      ...         ...        ...                      ...            ...   
1312       1           0 2022-03-13  Wolverhampton Wanderers        Everton   
1313       0           0 2022-03-18  Wolverhampton Wanderers   Leeds United   
1314       1           0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
1315       0           0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
1316       0           0 2022-04-24  Wolverhampton Wanderers        Burnley   

     result  
55        D  
56        W  
57        W  
58        W  
59        W  
...     ...  
1312      W  
1313      L  
1314      W  
1315      L  
1316      L  

[276 rows x 6 columns]

In [49]:
class MissingDict(dict):
  __missing__ = lambda self, key: key

map_values = {
    "Brighton and Hove Albion": "Brighton",
    "Manchester United": "Manchester Utd",
    "Newcastle United": "Newcastle Utd",
    "Tottenham Hotspur": "Tottenham",
    "West Ham United": "West Ham",
    "Wolverhampton Wanderers": "Wolves"
}

mapping = MissingDict(**map_values)

In [50]:
mapping["West Ham United"]

'West Ham'

In [51]:
combined["new_team"] = combined["team"].map(mapping)

In [52]:
combined

actual  prediction       date                     team       opponent  \
55         0           0 2022-01-23                  Arsenal        Burnley   
56         1           0 2022-02-10                  Arsenal         Wolves   
57         1           0 2022-02-19                  Arsenal      Brentford   
58         1           1 2022-02-24                  Arsenal         Wolves   
59         1           1 2022-03-06                  Arsenal        Watford   
...      ...         ...        ...                      ...            ...   
1312       1           0 2022-03-13  Wolverhampton Wanderers        Everton   
1313       0           0 2022-03-18  Wolverhampton Wanderers   Leeds United   
1314       1           0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
1315       0           0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
1316       0           0 2022-04-24  Wolverhampton Wanderers        Burnley   

     result new_team  
55        D  Arsenal  
56        W  Arsenal  
57        W  Arsenal  
58        W  Arsenal  
59        W  Arsenal  
...     ...      ...  
1312      W   Wolves  
1313      L   Wolves  
1314      W   Wolves  
1315      L   Wolves  
1316      L   Wolves  

[276 rows x 7 columns]

In [53]:
merged = combined.merge(combined, left_on = ["date", "new_team"], right_on = ["date", "opponent"])

In [54]:
merged

actual_x  prediction_x       date                   team_x  \
0           0             0 2022-01-23                  Arsenal   
1           1             0 2022-02-10                  Arsenal   
2           1             0 2022-02-19                  Arsenal   
3           1             1 2022-02-24                  Arsenal   
4           1             1 2022-03-06                  Arsenal   
..        ...           ...        ...                      ...   
257         1             0 2022-03-13  Wolverhampton Wanderers   
258         0             0 2022-03-18  Wolverhampton Wanderers   
259         1             0 2022-04-02  Wolverhampton Wanderers   
260         0             0 2022-04-08  Wolverhampton Wanderers   
261         0             0 2022-04-24  Wolverhampton Wanderers   

        opponent_x result_x new_team_x  actual_y  prediction_y  \
0          Burnley        D    Arsenal         0             0   
1           Wolves        W    Arsenal         0             0   
2        Brentford        W    Arsenal         0             0   
3           Wolves        W    Arsenal         0             0   
4          Watford        W    Arsenal         0             0   
..             ...      ...        ...       ...           ...   
257        Everton        W     Wolves         0             0   
258   Leeds United        L     Wolves         1             0   
259    Aston Villa        W     Wolves         0             0   
260  Newcastle Utd        L     Wolves         1             0   
261        Burnley        L     Wolves         1             0   

                      team_y opponent_y result_y     new_team_y  
0                    Burnley    Arsenal        D        Burnley  
1    Wolverhampton Wanderers    Arsenal        L         Wolves  
2                  Brentford    Arsenal        L      Brentford  
3    Wolverhampton Wanderers    Arsenal        L         Wolves  
4                    Watford    Arsenal        L        Watford  
..                       ...        ...      ...            ...  
257                  Everton     Wolves        L        Everton  
258             Leeds United     Wolves        W   Leeds United  
259              Aston Villa     Wolves        L    Aston Villa  
260         Newcastle United     Wolves        W  Newcastle Utd  
261                  Burnley     Wolves        W        Burnley  

[262 rows x 13 columns]

In [55]:
merged[(merged["prediction_x"] == 1) & (merged["prediction_y"] == 0)]["actual_x"].value_counts()

actual_x
1    27
0    13
Name: count, dtype: int64

In [56]:
27/40

0.675